### DISCUSSION TITLE
Report failures should be more obvious and fixable from the Web GUI

### DISCUSSION THREAD

**Discussion by JCWasmx86** [netbox-community/netbox#13664] [Ideas]
If you have e.g. accidentally uploaded a report without test methods, the entire reports page is unusable, due to not giving you the possibility of removing the erroneous report without logging in to the server running Netbox, etc. It would be better, if you could simply remove this report with one click.

For example, you could do it like that:
![image](https://github.com/netbox-community/netbox/assets/63594136/9c8094c5-5c5b-4dee-865f-8ddf03c076a5)

Another issue is, that during the upload of a report with e.g. syntax errors, you just get "Could not load reports from b.py", albeit in the log there is already `Unable to load report module b.py, exception: 'break' outside loop (b.py, line 6)` for e.g. this `b.py`:
```
from extras.reports import Report


class DeviceConnectionsReport1(Report):
    description = "Validate the minimum physical connections for each device"
    break
```
It would be useful to propagate this error to the UI. For example like this:

![image](https://github.com/netbox-community/netbox/assets/63594136/04ac5b37-3387-4b04-9929-f9f51cbf858b)

The code is here for playing around: https://github.com/netbox-community/netbox/compare/develop...JCWasmx86:netbox:develop

*Nothing really concrete, so I created a discussion, hope that's the correct thing to do*

**Comment by candlerb** [comment]
Already raised as #12842 (for reports) and #12766 (for scripts).

Reports were actually more broken than this - they caused the whole UI to raise an exception - and the pull which closed #12842 fixed that.  But now reports and scripts are both the same, in that they just say "Could not load" without any more info.

For now, my workaround is to add these lines temporarily to the top of the script or report:

```
import django, os, sys
sys.path.append('/opt/netbox/netbox')
os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'netbox.settings')
django.setup()
```

and then run from the command line:

```
/opt/netbox/venv/bin/python myreport.py
```

That should show you the error which prevents the script loading. When you've fixed it, comment out or remove the temporary lines.

**Reply by JCWasmx86** [reply]
I observed the changes to the handling of invalid reports - thanks for that. The workaround looks useful, but the reports page still has the issue that uploading e.g. an report without test methods by example causes the entire extras/reports page to be useless => That's bad

So would you consider - if my proposals are fleshed out a bit more if needed - taking patches to improve the behavior? (And I think it would improve the UX) Or do you consider the current state, that you have to drop to the CLI to fix some stuff to be good enough?

**Reply by candlerb** [reply]
> the issue that uploading e.g. an report without test methods by example causes the entire extras/reports page to be useless

If you can give a reproducer then raise an issue for it.

> taking patches to improve the behavior?

I don't see why not: first you need an accepted issue for the bug, and then you can submit a patch for it, and then the project maintainers will decide whether or not to accept it.

**Reply by JCWasmx86** [reply]
I've submitted #13666 for the broken report page, if there are no test reports. Regarding the propagation of syntax errors to the GUI, I can open a issue for that one too, if wished

**Comment by llamafilm** [comment]
How do you see the error message in the log?  I've checked `journalctl` for netbox and netbox-rq and not much appears there.  I did find it by enabling Django debug log as below, but it's extremely verbose and hard to read.  (That log file is 23GB after just a few days.)
```
LOGGING = {
    'version': 1,
    'disable_existing_loggers': False,
    'handlers': {
        'file': {
            'level': 'DEBUG',
            'class': 'logging.FileHandler',
            'filename': '/var/log/netbox.log',
        },
    },
    'loggers': {
        'django': {
            'handlers': ['file'],
            'level': 'INFO',
        },
        'netbox': {
            'handlers': ['file'],
            'level': 'DEBUG',
        },
    },
}
```

**Reply by JCWasmx86** [reply]
Personally I just wrap the test_* methods in a try-except block (I'm not the admin of the netbox instance I use)
